File: sdks/node/src/seed/resources/ota.ts
/** OTA config + check-now — Phase 1 resources. */

import type { CallOptions } from "../callOptions.js";

export interface OtaConfig extends Record<string, unknown> {
  channel?: string;
  auto_update?: boolean;
  check_interval_secs?: number;
}

export interface OtaCheckResponse extends Record<string, unknown> {
  update_available?: boolean;
  current_version?: string;
  latest_version?: string;
  checked_at?: string;
}

type RequestFn = <T>(
  method: string,
  path: string,
  opts?: CallOptions & {
    body?: unknown;
    idempotent?: boolean;
  },
) => Promise<T>;

export interface OtaResource {
  /** GET /api/v1/ota/config — WiFi-read allowlist. */
  config(opts?: CallOptions): Promise<OtaConfig>;
  /** POST /api/v1/ota/check-now — idempotent probe; safe to retry. */
  checkNow(opts?: CallOptions): Promise<OtaCheckResponse>;
}

export function makeOtaResource(request: RequestFn): OtaResource {
  return {
    config: (opts) =>
      request<OtaConfig>("GET", "/api/v1/ota/config", {
        idempotent: true,
        ...(opts ?? {}),
      }),

    checkNow: (opts) =>
      request<OtaCheckResponse>("POST", "/api/v1/ota/check-now", {
        idempotent: true, // the seed merely re-checks; no destructive effect
        ...(opts ?? {}),
      }),
  };
}

File: sdks/rust/src/seed/resources/ota.rs
//! OTA resource — `/api/v1/ota/*`.

use crate::error::Error;
use crate::seed::client::SeedClient;
use crate::seed::config::CallOptions;
use crate::seed::models::{OtaCheckNowAck, OtaConfig};

/// OTA endpoints.
pub struct OtaResource<'c> {
    pub(crate) client: &'c SeedClient,
}

impl<'c> OtaResource<'c> {
    /// `GET /api/v1/ota/config` (allowlisted read).
    pub async fn config(&self) -> Result<OtaConfig, Error> {
        self.client.request_get("/ota/config").await
    }

    /// [`Self::config`] with per-call [`CallOptions`] overrides.
    pub async fn config_with(&self, opts: CallOptions) -> Result<OtaConfig, Error> {
        self.client.request_get_opts("/ota/config", &opts).await
    }

    /// `POST /api/v1/ota/check-now` — v0.20.0+. Requires pairing token
    /// (seed returns 403 otherwise). Not idempotent server-side but
    /// repeat-safe — the seed dedupes internally — so mark idempotent.
    pub async fn check_now(&self) -> Result<OtaCheckNowAck, Error> {
        self.client
            .request_post::<OtaCheckNowAck, _>("/ota/check-now", &serde_json::json!({}), true)
            .await
    }

    /// [`Self::check_now`] with per-call [`CallOptions`] overrides.
    pub async fn check_now_with(&self, opts: CallOptions) -> Result<OtaCheckNowAck, Error> {
        self.client
            .request_post_opts::<OtaCheckNowAck, _>(
                "/ota/check-now",
                &serde_json::json!({}),
                true,
                &opts,
            )
            .await
    }
}
